Added the actual post since I made a mistake beforehand. (#14)
BiAPoL/blog: ryan_savill/principal_feature_analysis/readme.md

If you have not heard of PCA before I would advise you to check out [this video](https://www.youtube.com/watch?v=FgakZw6K1QQ). If you (like me just a few days ago) have an understanding of the principle of PCA but don't understand the math behind it I'd advise you to take some time to grasp the concepts. I found a great series of videos that explain all necessary parts of the puzzle that is PCA. Each video is just about 10 minutes long and explained quite well, even if your math lectures have been a few years ago!

- [Vector Projections](https://www.youtube.com/watch?v=X78tLBY3BMk)
- [Eigenvectors and Eigenvalues](https://www.youtube.com/watch?v=glaiP222JWA)
- [Derivative of a Matrix](https://www.youtube.com/watch?v=e73033jZTCI)
- [Lagrange Multipliers](https://www.youtube.com/watch?v=6oZT72-nnyI)
- [Covariance Matrix](https://www.youtube.com/watch?v=152tSYtiQbw)
- [PCA](https://www.youtube.com/watch?v=dhK8nbtii6I) (finally!)

Once you have grasped these concepts this explanation should be a breeze!

> What principal feature analysis does is take advantage of some of the transformations in PCA to find out which features explain the variance. The first steps in PFA are to create the covariance matrix and find the orthonormal eigenvectors and sort them by their eigenvectors, creating a matrix A. This in effect is what PCA does as well since the principal components are the eigenvectors of the covariance matrix. We can output them in [scikit-learn](https://scikit-learn.org/stable/) with the command '.components_'. 

Reviewer: ```suggestion
Principal feature analysis takes advantage of some of the transformations in PCA to find out which features explain the variance. The first steps in PFA are to create the covariance matrix and find the orthonormal eigenvectors and sort them by their eigenvalues, creating a matrix A. This in effect is what PCA does as well since the principal components are the eigenvectors of the covariance matrix. We can output them in [scikit-learn](https://scikit-learn.org/stable/) with the command '.components_'. 
```

```python
# importing everything we might need and more:
from sklearn import decomposition
import tribolium_clustering as tc
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from collections import defaultdict
from sklearn.metrics.pairwise import euclidean_distances
from sklearn.preprocessing import StandardScaler

# function for labelling eigenvectors so it is readable:
def pca_comp_inf_dataframe(pca_object,regprops_df,n_components):
    from sklearn.decomposition import PCA
    import pandas as pd
    import numpy as np
    features = regprops_df.keys()
    component_names = ['PC{}'.format(i+1) for i in range(n_components)]
    comps = pca_object.components_
    df_comps = pd.DataFrame(comps,index = component_names, columns = features)
    
    return df_comps

# loading our test dataset

> location_prefix = 'C://Users//ryans//OneDrive//Documents//'

Reviewer: ```suggestion
location_prefix = 'data/folder/'
```

> # location_prefix = D://Uni//MSTER TUD//'

Reviewer: ```suggestion

```

timepoint = 9
regionpropspath = location_prefix + 'Master Thesis//First Coding Tries//regionprops_all_timepoints_lund//regprops t{}.csv'.format(timepoint)
regprops = tc.readcsv_as_cl_input(regionpropspath)

# redefinition of our dataset
X = regprops

# standardscaling the dataset is a prerequisite for PCA
sc = StandardScaler()
X = sc.fit_transform(X)

# fitting our PCA
pca = PCA().fit(X)

# getting our orthonormal eigenvectors:
A = pca.components_.T

# labelling that matrix for ease of use
A_readable = pca_comp_inf_dataframe(pca,regprops,len(regprops.keys())).transpose()

A_readable
```

>     No Predictions in Regionprops of C://Users//ryans//OneDrive//Documents//Master Thesis//First Coding Tries//regionprops_all_timepoints_lund//regprops t9.csv

Reviewer: ```suggestion
    No Predictions in Regionprops of data/folder/Master Thesis/First Coding Tries/regionprops_all_timepoints_lund/regprops t9.csv
```

      <td>0.212046</td>
      <td>0.076844</td>
      <td>-0.308744</td>
      <td>0.228016</td>
      <td>-0.034887</td>
      <td>-0.121603</td>
      <td>-0.049914</td>
      <td>-0.068178</td>
      <td>0.120108</td>
      <td>-0.802665</td>
    </tr>
    <tr>
      <th>avg distance of 4 closest points</th>
      <td>0.316944</td>
      <td>0.023361</td>
      <td>0.033518</td>
      <td>-0.059729</td>
      <td>-0.078356</td>
      <td>-0.068524</td>
      <td>0.204538</td>
      <td>0.092240</td>
      <td>-0.253916</td>
      <td>0.188049</td>
      <td>-0.088972</td>
      <td>-0.156060</td>
      <td>-0.072032</td>
      <td>-0.097133</td>
      <td>0.660690</td>
      <td>0.499889</td>
    </tr>
    <tr>
      <th>touching neighbor count</th>
      <td>-0.190434</td>
      <td>-0.078438</td>
      <td>-0.284396</td>
      <td>-0.437245</td>
      <td>-0.339335</td>
      <td>0.594522</td>
      <td>-0.243295</td>
      <td>0.334427</td>
      <td>-0.201843</td>
      <td>0.052019</td>
      <td>-0.024418</td>
      <td>0.025692</td>
      <td>0.012726</td>
      <td>0.013417</td>
      <td>0.028237</td>
      <td>0.006079</td>
    </tr>
  </tbody>
</table>
</div>

> Now we have a massive matrix of our eigenvectors sorted by eigenvalues, but since every feature has equal influence on the eigenvectors we need to take a subset of matrix A called A_q. 'q' represents the number of eigenvectors we take into account and the value of q will influence how much variance we can explain with the subset of the principal components. We can turn this question around and set a threshold of how much variance should at least be explained and then find the corresponding subset A-q of matrix A:

Reviewer: ```suggestion
Now we have a massive matrix of our eigenvectors sorted by eigenvalues, but since every feature has equal influence on the eigenvectors we need to take a subset of matrix `A` called `A_q`. `q` represents the number of eigenvectors we take into account and the value of `q` will influence how much variance we can explain with the subset of the principal components. We can turn this question around and set a threshold of how much variance should at least be explained and then find the corresponding subset `A-q` of matrix `A`:
```

      <td>-0.078562</td>
      <td>-0.095429</td>
    </tr>
    <tr>
      <th>min_intensity</th>
      <td>-0.248677</td>
      <td>-0.277709</td>
      <td>-0.003527</td>
      <td>-0.151868</td>
      <td>-0.443942</td>
    </tr>
    <tr>
      <th>image_stdev</th>
      <td>-0.124373</td>
      <td>0.597732</td>
      <td>-0.026935</td>
      <td>0.059831</td>
      <td>0.275986</td>
    </tr>
    <tr>
      <th>avg distance of 2 closest points</th>
      <td>0.312403</td>
      <td>0.040115</td>
      <td>0.025366</td>
      <td>-0.113510</td>
      <td>-0.083929</td>
    </tr>
    <tr>
      <th>avg distance of 3 closest points</th>
      <td>0.315714</td>
      <td>0.031638</td>
      <td>0.028109</td>
      <td>-0.084834</td>
      <td>-0.084732</td>
    </tr>
    <tr>
      <th>avg distance of 4 closest points</th>
      <td>0.316944</td>
      <td>0.023361</td>
      <td>0.033518</td>
      <td>-0.059729</td>
      <td>-0.078356</td>
    </tr>
    <tr>
      <th>touching neighbor count</th>
      <td>-0.190434</td>
      <td>-0.078438</td>
      <td>-0.284396</td>
      <td>-0.437245</td>
      <td>-0.339335</td>
    </tr>
  </tbody>
</table>
</div>

> If you have read the paper or are following along with the paper we have just completed step 3 of the explained algorithm. What is done now is that we perform clustering on the matrix A_q with K-means clustering. Essentially what we are trying to do is find clusters of features as features that are clustered in this matrix have similar influences on the principal components and are thus correlating and not telling us much. This way we can just choose the feature that is closest to the cluster centre and it will be chosen as a principal feature. 

Reviewer: ```suggestion
If you have read the paper or are following along with the paper we have just completed step 3 of the explained algorithm. What is done now is that we perform clustering on the matrix A_q with K-means clustering. Essentially, we are trying to find clusters of features as features that are clustered in this matrix have similar influences on the principal components and are thus correlating and not telling us much. This way we can just choose the feature that is closest to the cluster centre and it will be chosen as a principal feature. 
```

> We can see that the choice of the cluster-number will determine how many features we choose. Y. Lu et al. recommend choosing the cluster-number: p, as: p >= q because we can't be completely sure that the variance that we chose is explained when p = q. We can just opt for a slightly larger p (arbitrary choice of 2  larger than q) but there is room for improvement in terms of an automated choice here OR it could be a parameter that we can implement later.

Reviewer: ```suggestion
We can see that the choice of the cluster-number will determine how many features we choose. Y. Lu et al. recommend choosing the cluster-number: p, as: `p >= q` because we can't be completely sure that the variance that we chose is explained when `p = q`. We can just opt for a slightly larger `p` (arbitrary choice of 2  larger than `q`) but there is room for improvement in terms of an automated choice here OR it could be a parameter that we can implement later.
```

> Since K-means is implemented well in scikit-learn this procedure is just a couple of lines:

Reviewer: ```suggestion
Since K-means-clustering is implemented well in [scikit-learn](https://scikit-learn.org/stable/modules/clustering.html) this procedure is just a couple of lines:
```

# iterating through each feature (i) and it's cluster (c)
for i, c in enumerate(clusters):
    
    # measuring the distance of each feature in A_q ([A_q[i, :]])
    # and the cluster centers [cluster_centers[c, :]]
    dist = euclidean_distances([A_q[i, :]], [cluster_centers[c, :]])[0][0]
    
    # each cluster is a key in the dictionary and the distance of each feature
    # belonging to that cluster is saved under this key as a tupule (feature_idx, distance)
    dists[c].append((i, dist))

dists
```

    defaultdict(list,
                {2: [(0, 0.039527541499107074),
                  (4, 0.22202861661398224),
                  (5, 0.1470198671974888),
                  (6, 0.13180725994276707),
                  (12, 0.09346378648078914),
                  (13, 0.06610212866152891),
                  (14, 0.04944098356286365)],
                 5: [(1, 0.0)],
                 6: [(2, 0.0)],
                 1: [(3, 0.0)],
                 4: [(7, 0.0)],
                 0: [(8, 0.217150845350435),
                  (10, 0.2117088702247021),
                  (15, 0.29315857787851873)],
                 3: [(9, 0.21819421392343466), (11, 0.21819421392343452)]})

> In our dists dictionary eah cluster has a list of tupules associated with it. Each tuple contains:

Reviewer: ```suggestion
In our `dists` dictionary each cluster has a list of tuples associated with it. Each tuple contains:
```

(feature index, distance to cluster center)

> We can see that for the clusters with only one tupule associated (5,6,1,4), the features all have distance 0 since the cluster center IS the feature. For all other clusters we still have to determine which feature is closest to the cluster center. Once the features have been determined we can transform our original dataset to only include these features:

Reviewer: ```suggestion
We can see that for the clusters with only one tuple associated `(5,6,1,4)`, the features all have distance 0 since the cluster center IS the feature. For all other clusters we still have to determine which feature is closest to the cluster center. Once the features have been determined we can transform our original dataset to only include these features:
```

                if j >= self.explained_var:
                    q = i
                    break
                    
        A_q = pca.components_.T[:,:q]
        
        clusternumber = min([q + self.diff_n_features, X.shape[1]])
        
        kmeans = KMeans(n_clusters= clusternumber).fit(A_q)
        clusters = kmeans.predict(A_q)
        cluster_centers = kmeans.cluster_centers_

        dists = defaultdict(list)
        for i, c in enumerate(clusters):
            dist = euclidean_distances([A_q[i, :]], [cluster_centers[c, :]])[0][0]
            dists[c].append((i, dist))

        self.indices_ = [sorted(f, key=lambda x: x[1])[0][0] for f in dists.values()]
        self.features_ = X[:, self.indices_]
        
        return X[:, self.indices_]
    
    def transform(self, X):
        return X[:, self.indices_]

# Testing
pfa = PFA(diff_n_features=1, explained_var= 0.95)
pfa.fit_transform(regprops)

featurekeys = [regprops.keys().tolist()[i] for i in pfa.indices_]
featurekeys
```

    ['area',
     'centroid-0',
     'centroid-1',
     'solidity',
     'mean_intensity',
     'image_stdev',
     'touching neighbor count']

> This is a modification of the provided code in that it tries to force you to choose a subset. If you don't manually say what the subset is, it will get a subset that explains 95% of the variance or the amount of variance that you determine. As described in the paper the number of clusters (and with that the number of features) should be larger than the subset of eigenvectors chosen. This is implemented by not determining the number of features but only setting how many more features than the subset should be chosen. Apart from this a transform function was added as well as a fit_transform function, similarly to other scikit learn functions. (DISCLAIMER FOR CODE USAGE?)

Reviewer: ```suggestion
This is a modification of the provided code in that it tries to force you to choose a subset. If you don't manually say what the subset is, it will get a subset that explains 95% of the variance or the amount of variance that you determine. As described in the paper, the number of clusters (and with that the number of features) should be larger than the subset of eigenvectors chosen. This is implemented by not determining the number of features but only setting how many more features than the subset should be chosen. Apart from this a transform function was added as well as a fit_transform function, similarly to other scikit learn functions.
